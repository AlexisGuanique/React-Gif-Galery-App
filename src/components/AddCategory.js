import React, { useState } from 'react'
import PropTypes from 'prop-types';


// Creamos componente para aÃ±adir una categoria
export const AddCategory = ( { setCategories } ) => {
    
    // Definimos el estado inicial de nuestro value
    const [InputValue, setInputValue] = useState('');


    // Extraemos el valor a guardar
    const handlInputChange = (e) => {
        // console.log(e.target.value);
        setInputValue(e.target.value);

        // console.log('HandleInputChange llamado')
    }


    // De esta manera evitamos que refresque el formulario de la manera que viene por defecto
    const handleSubmit = (e) => {
        e.preventDefault()

        // console.log('handleSubmit')

        // Validamos que el valor ingresado sea una palabra mayor de 2 letras
        if ( InputValue.trim().length > 1) {

            setCategories( cats => [ InputValue, ...cats ]);
            setInputValue('');
        }
    };
  

    // Lo que renderiza nuestro componente
    return (
        <form onSubmit={ handleSubmit }>
            <p>{ InputValue }</p>
            <input 
                type="text"
                value={ InputValue }
                onChange= { handlInputChange }         
            />
        </form>
  )
};

// De esta manera decimos que la Prop setCategories es requerida para utilizar nuestro componente AddCategory
AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
};
