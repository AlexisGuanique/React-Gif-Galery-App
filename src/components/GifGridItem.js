import React from 'react';
import { PropTypes } from 'prop-types';

export const GifGridItem = ( {title, url } ) => {

    // console.log( id, title, url );

    return (
        <div className='card animate__animated animate__fadeIn'>
            <img src={ url } alt={ title }/>
            <p className='title'>{ title }</p>
        </div>
    )
};


// Configuracion de los PropTypes, de esta manera indicamos los lineamientos para trabajar con nuestro componente
GifGridItem.protoType = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
};
