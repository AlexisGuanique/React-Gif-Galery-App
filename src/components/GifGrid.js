import React from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';


// Componente que consume la api_key y muestra los gif
export const GifGrid = ( { category } ) => {


    // Utilizamos nuestro custom hook
    const { data:images, loading } = useFetchGifs( category );

    console.log(images);

    return (
        <>
            <h2 className='animate__animated animate__fadeIn'> { category } </h2>

            { loading && <p className='animate__animated animate__flash'>Cargando...</p> }
            
            <div className='card-grid'>
                {
                    images.map( img => (
                        <GifGridItem
                            key={ img.id } 
                            { ...img }
                        />
                    ))
                }
            </div>
        </>
    )
};

