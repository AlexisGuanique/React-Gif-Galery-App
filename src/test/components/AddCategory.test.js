import React from "react";
import '@testing-library/jest-dom';
import { AddCategory } from "../../components/AddCategory";
import { shallow } from 'enzyme';

describe('Pruebas en <AddCategory />', () => {

    // Con el jest.fn() podemos saber cuando fue llamada, como fue llamada, etc, esta funcion setCategories
    const setCategories = jest.fn() ;
    let wrapper = shallow( <AddCategory setCategories={ setCategories } /> );


    // Esta funcion nos ayuda a reuniciar algun valor
    beforeEach(()=>{
        // clearAllMocks sirve para realizar una limpieza de algo
        jest.clearAllMocks()
        wrapper = shallow( <AddCategory setCategories={ setCategories } /> );
    })

    test('Debe de mostrarse cottectamente', () => {

        expect( wrapper ).toMatchSnapshot();

    });


    test('Debe de cambiar la caja de texto', () => {

        const input = wrapper.find('input');
        const value = 'Hola Mundo';

        input.simulate('change', { target: { value } });

        expect( wrapper.find('p').text().trim() ).toBe( value );

    });

    test('No debe de postear la informacion con submit',() => {

        // Le enviamos el preventDefault como argumento al simulate
        wrapper.find('form').simulate('submit', { preventDefault(){} })

        expect( setCategories ).not.toHaveBeenCalled();

    });

    test('Debe de llamar setCategories y limpiar la caja de texto', () => {

        const value = 'Hola Mundo'
        

        // 1. simular el inputChance

        wrapper.find('input').simulate('change', { target: { value }});


        // 2. simular el submit

        wrapper.find('form').simulate('submit', { preventDefault(){}} );
        
        
        // 3. setCategories se debe de haber llamado

        expect( setCategories ).toHaveBeenCalled();

        
        // 4. el valor del input debe de estar ''

        expect( wrapper.find('input').prop( 'value' ) ).toBe( '' );
    })


})