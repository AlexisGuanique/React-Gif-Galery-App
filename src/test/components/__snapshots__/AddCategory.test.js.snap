// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddCategory /> Debe de mostrarse cottectamente 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
